### PR TITLE
use espree to parse the programs to enable JSX support and more ES6 features. also allows the user to replace the parser if needed.

### PR DESCRIPTION
**Pull request by millermedeiros**
maybe we should add a way to toggle the parser from the config files as well but using [espree](https://github.com/eslint/espree) by default will be a good thing since the development is way faster than esprima and it already supports JSX and ES6 features out of the box.

@royriojas will like this feature since we will be able to deprecate his [esformatter-jsx-ignore](https://github.com/royriojas/esformatter-jsx-ignore) plugin. the main benefit is that we can have other plugins that use `stringAfter` without causing conflicts (if a plugin that doesn't understand JSX executes `stringAfter` after the `esformatter-jsx-ignore` it will fail)

---

see #192
see #242

### CODE REVIEW

**Comment by royriojas** [comment]
Nice! yes, I'm all for deprecating the `esformatter-jsx-ignore`. are u planning to also support formatting for the jsx syntax that way we can deprecate the other `esformatter-jsx`? that would be great!

**Comment by royriojas** [comment]
Ah never mind!, I just saw the other comment, where say you won't add support on esformatter.

**Comment by millermedeiros** [comment]
main reason why I decided to use espree was because it supports more ES6 features, JSX was just a bonus.

not sure if we should enable jsx by default or let the `esformatter-jsx` plugin enable it -- if we enable it by default we might need to keep supporting it forever; if the plugin enables it, we might switch the parser in the future while `esformatter-jsx` can use `espree` or `esprima-fb` if needed.

we should definitely add a way to set the parser through the config file, that will allow plugins to override the parser and the parser settings.

**Comment by millermedeiros** [comment]
and a comment explaining why I think JSX formatting should be handled by plugins: https://github.com/millermedeiros/esformatter/issues/242#issuecomment-87947383

**Comment by millermedeiros** [comment]
I rebased it and planning to land it later this week with JSX enabled by default. not going to add support for custom parsers on the config file for now, maybe open a new feature request in case someone really think this would be useful. (config file is already too chaotic)

File: lib/format.js [hunk 1]
```diff
@@ -49,8 +83,10 @@ function getShebang(str) {
 
 function doFormat(str) {
   str = plugins.stringBefore(str);
-  var ast = rocambole.parse(str);
-  transform(ast, transform.BYPASS_OPTIONS);
+  // allows user to override the parser
+  rocambole.parseFn = exports.parseFn;
+  rocambole.parseContext = exports.parseContext;
+  var ast = rocambole.parse(str, exports.parseOptions);  transform(ast, transform.BYPASS_OPTIONS);
```
Comment (jzaefferer) [inline lib/format.js]: Missing linebreak

File: package.json [hunk 2]
```diff
@@ -40,19 +40,19 @@
...
   "dependencies": {
     "debug": "^0.7.4",
+    "espree": "^1.12.3",
     "glob": "^5.0.3",
     "mout": ">=0.9 <2.0",
     "npm-run": "^1.1.1",
     "optimist": "~0.6.0",
     "resolve": "^1.1.5",
-    "rocambole": ">=0.5 <2.0",
+    "rocambole": ">=0.6 <2.0",
```
Comment (jzaefferer) [inline package.json]: Why the giant range?

File: package.json [hunk 3]
```diff
@@ -40,19 +40,19 @@
...
     "esformatter-pipe-test": "file:test/pipe",
     "esformatter-test-plugin": "file:test/plugin",
-    "esprima": "^2.0.0",
     "jshint": "~2.3.0",
     "mocha": "https://github.com/millermedeiros/mocha/tarball/latest",
     "mockery": "^1.4.0"
   },
   "dependencies": {
     "debug": "^0.7.4",
+    "espree": "^1.12.3",
```
Comment (jzaefferer) [inline package.json]: How come that you're changing a devDependency (esprima) to a regular dependency (espree)?

**Comment by jzaefferer** [comment]
What other changes are need to support jsx? Is it just a matter of having the parse support it, or will you need new hooks as well? If the latter, enabling it by default might be a bad idea, since it'll entail lots more work.

File: package.json (hunk 3 above)
Comment (millermedeiros) [inline package.json]: before we used `esprima` on the tests to validate that the output was still valid JS. now that we depend on `espree` we need to use it as well on the validation.

File: package.json (hunk 2 above)
Comment (millermedeiros) [inline package.json]: If I ever change the `rocambole` API it will be after v2.0

**Comment by millermedeiros** [comment]
I don't have any plans to add JSX formatting into esformatter itself. I just want the parser to not throw any errors.

**Pull request closed without merging**